plugins/kotlin/gradle/gradle/tests/test/org/jetbrains/kotlin/idea/gradle/toolchain/GradleDaemonJvmCriteriaDownloadToolchainTest.kt

```diff
@@ -0,0 +1,103 @@
+// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+package org.jetbrains.kotlin.idea.gradle.toolchain
+
+import com.intellij.ide.projectWizard.generators.JdkDownloadService
+import com.intellij.openapi.projectRoots.impl.jdkDownloader.JdkDownloadTask
+import com.intellij.openapi.ui.Messages
+import com.intellij.openapi.ui.TestDialogManager
+import com.intellij.testFramework.fixtures.BasePlatformTestCase
+import com.intellij.testFramework.replaceService
+import com.intellij.util.ui.UIUtil
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+import org.mockito.kotlin.argumentCaptor
+import org.mockito.kotlin.times
+import java.nio.file.Files
+import kotlin.io.path.Path
+
+class GradleDaemonJvmCriteriaDownloadToolchainTest : BasePlatformTestCase() {
+
+    private lateinit var jdkDownloadService: JdkDownloadService
+
+    override fun setUp() {
+        super.setUp()
+        jdkDownloadService = mock(JdkDownloadService::class.java)
+        project.replaceService(JdkDownloadService::class.java, jdkDownloadService, testRootDisposable)
+    }
+
+    fun `test Given unsupported criteria When download matching JDK Then expected exception is thrown`() {
+        lateinit var exceptionMessage: String
+        TestDialogManager.setTestDialog({
+            exceptionMessage = it
+            Messages.OK
+        }, testRootDisposable)
+
+        createDaemonJvmPropertiesFile(0, "unknown vendor")
+        GradleDaemonJvmCriteriaDownloadToolchain.downloadJdkMatchingCriteria(project, project.basePath.orEmpty())
+
+        UIUtil.dispatchAllInvocationEvents()
+        assertTrue(exceptionMessage.startsWith("Failed to locate and download a matching toolchain"))
+    }
+
+    fun `test Given supported vendors When download matching JDK Then requested item contains expected values`() {
+        lateinit var lastError: Throwable
+        repeat(5) {
+            val result = runCatching {
+                listOf("Oracle", "Amazon", "BellSoft", "Azul", "SAP", "Eclipse", "IBM", "GraalVM", "JetBrains").forEachIndexed { index, vendor ->
+                    createDaemonJvmPropertiesFile(21, vendor)
+                    GradleDaemonJvmCriteriaDownloadToolchain.downloadJdkMatchingCriteria(project, project.basePath.orEmpty())
+
+                    assertJdkDownloader(21, vendor, index + 1)
+                }
+            }
+            if (result.isSuccess) return
+            lastError = result.exceptionOrNull()!!
+
+            if (lastError.message?.startsWith("Failed to locate and download a matching toolchain") == true) {
```

Comment:
### Flaky tests

Could you please mock the `JdkListDownloader` service? This allows stabilising tests and avoids test running in the repeat cycle.

P.S. If connecting to the remote server is necessary, then you need to configure JetBrains proxy to speed up and stabilise tests and avoid bans on the external services.

You can research examples starting from the [org.jetbrains.plugins.gradle.util.GradleEnvironment$Urls](https://github.com/JetBrains/intellij-community/blob/08256c880950ea3502cba89e8fbe8dea08d8c74b/plugins/gradle/src/org/jetbrains/plugins/gradle/util/GradleEnvironment.java#L22) class. But I recommend avoiding any internet connections in tests.

Reply:
This test was inspired by [JdkDownloaderIntegrationTest](https://github.com/JetBrains/intellij-community/blob/master/platform/lang-impl/testSources/com/intellij/openapi/projectRoots/impl/jdkDownloader/JdkDownloaderIntegrationTest.kt) so any change will be great to apply to both, potentially keeping it separate. In addition, this test do require to obtain the available JDKs to download from `https://download.jetbrains.com/jdk/feed/v1/jdks.json.xz` validating in this way more E2E flow since is easy that file got modified and then IDE isn't able to offer a matching criteria JDK to be downloaded. For this reason, mocking the `JdkListDownloader` will not make possible to spot those regressions.

Reply:
Moved away from it to use fake data instead, however, I leave this to you to ensure `https://download.jetbrains.com/jdk/feed/v1/jdks.json.xz` returns the expected format adding it as a backend check